### ISSUE TITLE
Question about notification

### ISSUE THREAD

**Issue by mincheolsung** [seL4/seL4#960]
Hi,

what happens if a thread sends a signal (seL4_Signal) on an endpoint before the other side is blocked (seL4_Wait)?
Is the signal pending such that the blocked thread wakes up immediately?

**Comment by malus-brandywine** [comment]
Hi,

seL4_Signal() is blocked until seL4_Wait() is called.

I would recommend seL4 Reference Manual: https://sel4.systems/Info/Docs/seL4-manual-latest.pdf

(Also, the designated communication channels for Q&A are the mailing list and Discourse.)

**Comment by gernotheiser** [comment]
seL4_Signal() never blocks, that's the whole point of Notifications. It's like a semaphore Signal operation.

**Comment by malus-brandywine** [comment]
Sorry about the earlier confusion. Using the occasion, I'd like to bring back the topic I brought before ( #644 ).
API Reference describes some API functions as wrappers of other functions, meanwhile the latter ones are wrappers themselves. So, `seL4_Signal()` is described as a wrapper for blocking `seL4_Send()`.

In fact, both of them are wrappers to system call of type "send", which defines the direction of information flow.
So, `seL4_Signal()` is a non-blocking version of sys.call type "send", `seL4_Send()` is a blocking version of it.
It brings three types of internal sys.calls: "send", "receive" and "yield".

**Comment by mincheolsung** [comment]
Hi,
I observed that seL4_Signal() is blocking before no receiving end called seL4_Wait().

**Comment by malus-brandywine** [comment]
Hi,
there can be a trap: if a waiting thread is of higher priority and never yields, then a low priority signalling thread will not run again.

**Comment by mincheolsung** [comment]
The initial task calls seL4_Signal(). I thought it has the highest priority (255).

**Comment by kent-mcleod** [comment]
`seL4_Signal()` is an alias for seL4_Send() (https://docs.sel4.systems/projects/sel4/api-doc.html#signal). seL4_Signal() is supposed to only be used on notifications where it will never block. If it is used on an endpoint, then it can block if there is no receiver waiting on the endpoint. seL4_NBSend() should be used instead of seL4_Send() if you want to attempt to send to an endpoint without blocking. For notifications, seL4_NBSend() and seL4_Send() have the same behavior as a thread cannot block on sending for a notification.

**Comment by kent-mcleod** [comment]
@mincheolsung are you using an endpoint or notification object? The issue description asks about endpoints while the title of the issue is about notifications.  

@malus-brandywine, I agree with your description for when seL4_Signal is called on a notification object capability.

When the invoked capability is for an endpoint, then `decodeInvocation` will handle the syscall with the `cap_endpoint_cap` path which can block.  So whether seL4_Signal is allowed to block or not is first dependent on whether the capability being invoked is a notification or endpoint. (As we agree that this because from the kernel's perspective it's just a seL4_SysSend call regardless of whether user called seL4_Send or seL4_Signal)
